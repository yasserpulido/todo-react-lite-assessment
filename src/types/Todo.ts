export interface Todo {
  id: string;
  value: string;
  isDone: boolean;
}
