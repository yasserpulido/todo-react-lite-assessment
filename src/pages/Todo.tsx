/** @jsxImportSource @emotion/react */
import React, { useState } from "react";
import NewTodo from "../components/todo/NewTodo";
import TodoList from "../components/todo/TodoList";
import { Todo as TodoType } from "../types/Todo";
import { css } from "@emotion/react";

const DUMMY_DATA: TodoType[] = [
  {
    id: "t01",
    value: "Aprender TypeScript",
    isDone: false,
  },
  {
    id: "t02",
    value: "Hacer Tarea",
    isDone: true,
  },
  {
    id: "t03",
    value: "Entregar Tarea",
    isDone: false,
  },
];

const todoCard = css`
  margin: 4rem auto;
  padding: 2rem 3rem 3rem;
  max-width: 500px;
  background-color: #fbc531;
  color: #000;
  box-shadow: 5px 5px 0px 0px rgb(0, 0, 0, 1);
  border: 2px solid #000;
  border-radius: 4px;
`;

const todoTitle = css`
  text-align: center;
`;

const showDoneButton = css({
  backgroundColor: "#f5f6fa",
  border: "1px solid #95a5a6",
  cursor: "pointer",
  padding: "6px 16px",
  borderRadius: "4px",
  marginBottom: "8px",
  textAlign: "right",
  "&:hover": {
    backgroundColor: "#dcdde1",
  },
});

const formWrap = css({
  textAlign: "right",
});

const Todo: React.FC = () => {
  const [todos, setTodos] = useState(DUMMY_DATA);
  const [showDone, setShowDone] = useState(false);

  const newTodoHandler = (todo: string): void => {
    const idGenerated = Math.floor(Math.random() * 100).toString();

    setTodos((prevState: TodoType[]) => {
      return [...prevState, { id: idGenerated, value: todo, isDone: false }];
    });
  };

  const doneToggleHandler = (id: string, checked: boolean): void => {
    const n = todos;

    n.forEach((value) => {
      if (value.id === id) {
        value.isDone = checked;
      }
    });

    setTodos(() => {
      return [...n];
    });
  };

  const editTodoHandler = (id: string, value: string) => {
    const updatedTodo = todos.map((todo) => {
      if (todo.id === id) {
        return { ...todo, value };
      }
      return todo;
    });

    setTodos(updatedTodo);
  };

  const showDoneHandler = () => {
    setShowDone(!showDone);
  };

  const todosFilter = showDone
    ? todos
    : todos.filter((todo) => todo.isDone === showDone);

  return (
    <div css={todoCard}>
      <h1 css={todoTitle}>LISTA DE TAREAS</h1>
      <div css={formWrap}>
        <button css={showDoneButton} type="button" onClick={showDoneHandler}>
          {!showDone ? "Mostrar Completadas" : "Quitar Completadas"}
        </button>
        <NewTodo newTodo={newTodoHandler} />
      </div>
      <TodoList
        todos={todosFilter}
        doneToggle={doneToggleHandler}
        editTodo={editTodoHandler}
      />
    </div>
  );
};

export default Todo;
