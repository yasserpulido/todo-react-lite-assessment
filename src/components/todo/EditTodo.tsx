/** @jsxImportSource @emotion/react */
import React, { useState } from "react";
import { BaseSyntheticEvent } from "../../types/Events";
import { Todo } from "../../types/Todo";
import { css } from "@emotion/react";

interface Props {
  todo: Todo;
  editTodo: (a: string, b: string) => void;
}

const editInput = css({
  backgroundColor: "#fbc531",
  border: "2px solid #000",
  borderRadius: "4px",
  padding: "6px",
  margin: "8px",
  minWidth: "200px",
  outline: "none",
});

const saveButton = css({
  backgroundColor: "#f5f6fa",
  border: "1px solid #95a5a6",
  cursor: "pointer",
  padding: "6px 16px",
  borderRadius: "4px",
  marginBottom: "8px",
  textAlign: "right",
  "&:hover": {
    backgroundColor: "#dcdde1",
  },
});

const EditTodo: React.FC<Props> = (props) => {
  const [todo, setTodo] = useState({ ...props.todo });

  const submitHandler = (event: BaseSyntheticEvent) => {
    event.preventDefault();

    if (todo.text.trim().length === 0) {
      return;
    }

    props.editTodo(todo.id, todo.text);
  };

  const todoHandler = (event: BaseSyntheticEvent) => {
    setTodo((prevState: Todo) => {
      return { ...prevState, text: event.target.value };
    });
  };

  return (
    <form onSubmit={submitHandler}>
      <input
        css={editInput}
        type="text"
        value={todo.text}
        onChange={todoHandler}
      />
      <button css={saveButton} type="submit">
        Guardar
      </button>
    </form>
  );
};

export default EditTodo;
