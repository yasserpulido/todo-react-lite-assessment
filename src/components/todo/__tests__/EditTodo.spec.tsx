import { fireEvent, render, screen } from "@testing-library/react";
import { Todo } from "../../../types/Todo";
import EditTodo from "../EditTodo";

describe("EditTodo", () => {
  const todo: Todo = {
    id: "t01",
    text: "Hacer una prueba",
    done: false,
  };

  test("Should EditTodo form to match Snapshot", () => {
    // Act
    const { container } = render(<EditTodo editTodo={() => {}} todo={todo} />);

    // Assert
    expect(container).toMatchSnapshot();
  });

  test("Should EditTodo form render", () => {
    // Act
    render(<EditTodo editTodo={() => {}} todo={todo} />);
    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    // Assert
    expect(form).toBeInTheDocument();
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute("type", "text");
    expect(input).toHaveValue(todo.text);
    expect(button.innerHTML).toBe("Guardar");
  });

  test("Should EditTodo form submit and call createTodo method", () => {
    // Arrange
    const id = "t01";
    const text = "Hacer una prueba";
    const editTodo = jest.fn();

    // Act
    render(<EditTodo editTodo={editTodo} todo={todo} />);
    const form = screen.getByRole("form");
    fireEvent.submit(form);

    // Assert
    expect(editTodo).toHaveBeenCalled();
    expect(editTodo).toHaveBeenCalledTimes(1);
    expect(editTodo).toHaveBeenCalledWith(id, text);
  });

  test("Should not EditTodo form submit", () => {
    // Arrange
    const editTodo = jest.fn();

    // Act
    render(<EditTodo editTodo={editTodo} todo={todo} />);
    const form = screen.getByRole("form");
    const input = screen.getByRole("textbox");

    fireEvent.input(input, { target: { value: "" } });
    fireEvent.submit(form);

    // Assert
    expect(editTodo).not.toHaveBeenCalled();
  });
});
