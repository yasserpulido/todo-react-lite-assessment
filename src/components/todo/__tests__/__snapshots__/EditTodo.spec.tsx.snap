// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTodo Should EditTodo form to match Snapshot 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      class="css-14oliof-EditTodo"
      type="text"
      value="Hacer una prueba"
    />
    <button
      class="css-1g7u9vd-EditTodo"
      type="submit"
    >
      Guardar
    </button>
  </form>
</div>
`;
